src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java

```diff
@@ -0,0 +1,82 @@
+package org.apache.sysml.runtime.instructions.cp;
+
+import org.apache.sysml.runtime.DMLRuntimeException;
+import org.apache.sysml.runtime.controlprogram.caching.FrameObject;
+import org.apache.sysml.runtime.controlprogram.caching.MatrixObject;
+import org.apache.sysml.runtime.controlprogram.context.ExecutionContext;
+import org.apache.sysml.runtime.matrix.data.MatrixBlock;
+import org.apache.sysml.runtime.matrix.operators.Operator;
+import org.apache.sysml.runtime.util.DataConverter;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Eval built-in function instruction
+ * Note: it supports only single matrix[double] output
+ */
+public class EvalBuiltinNaryCPInstruction extends BuiltinNaryCPInstruction {
+
+	public EvalBuiltinNaryCPInstruction(Operator op, String opcode, String istr, CPOperand output, CPOperand... inputs) {
+		super(op, opcode, istr, output, inputs);
+	}
+
+	@Override
+	public void processInstruction(ExecutionContext ec) throws DMLRuntimeException {
+		//1. get the namespace and func
+		CPOperand func = inputs[0];
+		String funcName = func.getName();
+		String namespace = null;
+		if (func.getName().contains("::")) {
+			String[] split = func.getName().split("::");
+			namespace = split[0];
+			String funcSignature = split[1];
+			if (funcSignature.matches("(.+)([(].*[)])")) {
+				// get funcName from a function call identifier
+				Pattern pattern = Pattern.compile("(.+)([(].*[)])");
+				Matcher matcher = pattern.matcher(func.getName());
+				if (matcher.matches()) {
+					funcName = matcher.group(0);
+				}
+			} else {
+				funcName = funcSignature;
+			}
+		}
+		// bound the inputs to avoiding being deleted after the function call
+		CPOperand[] boundInputs = Arrays.asList(inputs).subList(1, inputs.length).toArray(new CPOperand[]{});
+		ArrayList<String> boundOutputNames = new ArrayList<>();
+		boundOutputNames.add(output.getName());
+		ArrayList<String> boundInputNames = new ArrayList<>();
+		for (CPOperand input : boundInputs) {
+			boundInputNames.add(input.getName());
+		}
+
+		//2. copy the created output matrix
+		MatrixObject outputMO = new MatrixObject(ec.getMatrixObject(output.getName()));
```

Comment:
There is no need to create a copy or even obtain this output object.

Reply:
I think that the eval func always returns a matrix. Then I set the ouput type with Matrix. So I copy the created output matrix object to avoid overriding it when the func call returns a scalar object.

Reply:
Yes, you right, I forgot about the scalar case. Since the copy constructor creates a different ID maybe you alternatively, get the object and disable its cleanup with `mo.enableCleanup(false)`, which preserves the contents but might drop it from the symbol table.